jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/RabitTracker.scala

```diff
@@ -0,0 +1,158 @@
+  *   tracker.waitFor(6 hours)
+  * }}}
+  *
+  * @param numWorkers number of distributed workers, each corresponding to a Spark task.
+  * @param port The minimum port that the tracker binds to. If port is omitted, or given as None,
+  *             a random ephemeral port is chosen at runtime.
+  * @param maxPortTrials The maximum number of trials of socket binding, by sequentially
+  *        increasing port.
+  * @param workerConnectionTimeout The timeout for awaiting connections from worker nodes.
+  *        When used in Spark applications, because all Spark transformations are lazily executed,
+  *        the I/O time for loading RDDs from external sources (local disk, HDFS, S3 etc.) must be
+  *        taken in account. If the timeout value is too small, the Rabit workers will likely time
+  *        out before connecting, due to the overhead of loading RDDs.
+  *        Using a finite timeout is encouraged, as it prevents the tracker (thus the Spark driver
+  *        running it) from hanging indefinitely due to worker connection issues (e.g. firewall).
+  */
+class RabitTracker(numWorkers: Int, port: Option[Int] = None,
+                   maxPortTrials: Int = 1000,
+                   workerConnectionTimeout: Duration = 10 minutes) extends IRabitTracker {
+  import scala.collection.JavaConverters._
+
+  require(numWorkers >=1, "numWorkers must be greater than or equal to one (1).")
+
+  val system = ActorSystem.create("RabitTracker")
+  val handler = system.actorOf(RabitTrackerHandler.props(
+    numWorkers, maxPortTrials, workerConnectionTimeout), "Handler")
+  implicit val askTimeout: akka.util.Timeout = akka.util.Timeout(1 second)
+
+  val futureWorkerEnvs: Future[Map[String, String]] = Try(
+    Await.result(handler ? RabitTrackerHandler.RequestBoundFuture, 5 seconds)
+      .asInstanceOf[Future[Map[String, String]]]
+  ) match {
+    case Success(fut) => fut
+    case Failure(e) =>
+      val delayedFailure = Promise[Map[String, String]]()
+      delayedFailure.failure(e)
+      delayedFailure.future
+  }
+
+  // a future for XGBoost worker execution
+  val futureCompleted: Future[Int] = Try(
+    Await.result(handler ? RabitTrackerHandler.RequestCompletionFuture, 5 seconds)
+      .asInstanceOf[Future[Int]]
+  ) match {
+    case Success(fut) => fut
+    case Failure(_) =>
+      val delayedFailure = Promise[Int]()
+      // mark 0 worker as completed, will trigger AssertionError
+      // when call waitFor()
+      delayedFailure.success(0)
+      delayedFailure.future
+  }
+
+  /**
+    * Start the Rabit tracker.
+    * @param timeout timeout for the tracker to bind to an address.
+    * @return
+    */
+  def start(timeout: Duration = 30 seconds): Boolean = {
+    handler ? RabitTrackerHandler.StartTracker(
+      new InetSocketAddress(InetAddress.getLocalHost,
+        // randomly allocate an ephemeral port if port is not specified
+        port.getOrElse(new Random().nextInt(61000 - 32768) + 32768)
```

Comment:
The `RabitTrackerHandler` class will handle the sequential probing of available port numbers, if the initial random one is already occupied. See https://github.com/xydrolase/xgboost/blob/c6addeb7a19a0bbc4c62f22cde0294b062303f6a/jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/handler/RabitTrackerHandler.scala#L105